// primeng/bundles/primeng-virtualscroller.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('@angular/cdk/scrolling'),require('primeng/api'),exports, require('@angular/core'), require('@angular/common'), require('primeng/api'), require('@angular/cdk/scrolling')) :
    typeof define === 'function' && define.amd ? define('primeng/virtualscroller', ['@angular/core','@angular/common','@angular/cdk/scrolling','primeng/api','exports', '@angular/core', '@angular/common', 'primeng/api', '@angular/cdk/scrolling'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.common,global.ng.cdk.scrolling,global.primeng.api,(global.primeng = global.primeng || {}, global.primeng.virtualscroller = {}), global.ng.core, global.ng.common, global.primeng.api, global.ng.cdk.scrolling));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,ɵngcc3,exports, core, common, api, scrolling) { 
var _c0 = ["viewport"];
function VirtualScroller_div_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 9);
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementEnd();
} }
function VirtualScroller_ng_container_7_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
var _c1 = function (a0) { return { "height": a0 }; };
var _c2 = function (a0, a1, a2, a3, a4, a5, a6) { return { $implicit: a0, index: a1, count: a2, first: a3, last: a4, even: a5, odd: a6 }; };
function VirtualScroller_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "li", 10);
    ɵngcc0.ɵɵtemplate(2, VirtualScroller_ng_container_7_ng_container_2_Template, 1, 0, "ng-container", 11);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var item_r902 = ctx.$implicit;
    var i_r903 = ctx.index;
    var c_r904 = ctx.count;
    var f_r905 = ctx.first;
    var l_r906 = ctx.last;
    var e_r907 = ctx.even;
    var o_r908 = ctx.odd;
    var ctx_r900 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(3, _c1, ctx_r900.itemSize + "px"));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", item_r902 ? ctx_r900.itemTemplate : ctx_r900.loadingItemTemplate)("ngTemplateOutletContext", ɵngcc0.ɵɵpureFunction7(5, _c2, item_r902, i_r903, c_r904, f_r905, l_r906, e_r907, o_r908));
} }
function VirtualScroller_div_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 12);
    ɵngcc0.ɵɵprojection(1, 1);
    ɵngcc0.ɵɵelementEnd();
} }
var _c3 = [[["p-header"]], [["p-footer"]]];
var _c4 = ["p-header", "p-footer"];
'use strict';

    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __read = (this && this.__read) || function (o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    };
    var __spread = (this && this.__spread) || function () {
        for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
        return ar;
    };
    var VirtualScroller = /** @class */ (function () {
        function VirtualScroller(el) {
            this.el = el;
            this.cache = true;
            this.first = 0;
            this.trackBy = function (index, item) { return item; };
            this.onLazyLoad = new core.EventEmitter();
            this._totalRecords = 0;
            this.lazyValue = [];
            this.page = 0;
        }
        Object.defineProperty(VirtualScroller.prototype, "totalRecords", {
            get: function () {
                return this._totalRecords;
            },
            set: function (val) {
                this._totalRecords = val;
                this.lazyValue = Array.from({ length: this._totalRecords });
                this.onLazyLoad.emit(this.createLazyLoadMetadata());
                this.first = 0;
                this.scrollTo(0);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(VirtualScroller.prototype, "value", {
            get: function () {
                return this.lazy ? this.lazyValue : this._value;
            },
            set: function (val) {
                if (this.lazy) {
                    if (val) {
                        var arr = this.cache ? __spread(this.lazyValue) : Array.from({ length: this._totalRecords });
                        for (var i = this.first, j = 0; i < (this.first + this.rows); i++, j++) {
                            arr[i] = val[j];
                        }
                        this.lazyValue = arr;
                    }
                }
                else {
                    this._value = val;
                }
            },
            enumerable: true,
            configurable: true
        });
        VirtualScroller.prototype.ngAfterContentInit = function () {
            var _this = this;
            this.templates.forEach(function (item) {
                switch (item.getType()) {
                    case 'item':
                        _this.itemTemplate = item.template;
                        break;
                    case 'loadingItem':
                        _this.loadingItemTemplate = item.template;
                        break;
                    default:
                        _this.itemTemplate = item.template;
                        break;
                }
            });
        };
        VirtualScroller.prototype.onScrollIndexChange = function (index) {
            var p = Math.floor(index / this.rows);
            if (p !== this.page) {
                this.page = p;
                this.first = this.page * this.rows;
                this.onLazyLoad.emit(this.createLazyLoadMetadata());
            }
        };
        VirtualScroller.prototype.createLazyLoadMetadata = function () {
            return {
                first: this.first,
                rows: this.rows
            };
        };
        VirtualScroller.prototype.getBlockableElement = function () {
            return this.el.nativeElement.children[0];
        };
        VirtualScroller.prototype.scrollTo = function (index) {
            if (this.viewPortViewChild && this.viewPortViewChild['elementRef'] && this.viewPortViewChild['elementRef'].nativeElement) {
                this.viewPortViewChild['elementRef'].nativeElement.scrollTop = index * this.itemSize;
            }
        };
        VirtualScroller.ctorParameters = function () { return [
            { type: core.ElementRef }
        ]; };
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "itemSize", void 0);
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "style", void 0);
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "styleClass", void 0);
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "scrollHeight", void 0);
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "lazy", void 0);
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "cache", void 0);
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "rows", void 0);
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "first", void 0);
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "trackBy", void 0);
        __decorate([
            core.ContentChild(api.Header, { static: true })
        ], VirtualScroller.prototype, "header", void 0);
        __decorate([
            core.ContentChild(api.Footer, { static: true })
        ], VirtualScroller.prototype, "footer", void 0);
        __decorate([
            core.ContentChildren(api.PrimeTemplate)
        ], VirtualScroller.prototype, "templates", void 0);
        __decorate([
            core.ViewChild('viewport', { static: true })
        ], VirtualScroller.prototype, "viewPortViewChild", void 0);
        __decorate([
            core.Output()
        ], VirtualScroller.prototype, "onLazyLoad", void 0);
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "totalRecords", null);
        __decorate([
            core.Input()
        ], VirtualScroller.prototype, "value", null);
VirtualScroller.ɵfac = function VirtualScroller_Factory(t) { return new (t || VirtualScroller)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
VirtualScroller.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: VirtualScroller, selectors: [["p-virtualScroller"]], contentQueries: function VirtualScroller_ContentQueries(rf, ctx, dirIndex) { if (rf & 1) {
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, api.Header, true);
        ɵngcc0.ɵɵstaticContentQuery(dirIndex, api.Footer, true);
        ɵngcc0.ɵɵcontentQuery(dirIndex, api.PrimeTemplate, false);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.header = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.footer = _t.first);
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.templates = _t);
    } }, viewQuery: function VirtualScroller_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.viewPortViewChild = _t.first);
    } }, inputs: { cache: "cache", first: "first", trackBy: "trackBy", totalRecords: "totalRecords", value: "value", itemSize: "itemSize", style: "style", styleClass: "styleClass", scrollHeight: "scrollHeight", lazy: "lazy", rows: "rows" }, outputs: { onLazyLoad: "onLazyLoad" }, ngContentSelectors: _c4, decls: 9, vars: 12, consts: [[3, "ngClass", "ngStyle"], ["class", "ui-virtualscroller-header ui-widget-header ui-corner-top", 4, "ngIf"], [1, "ui-virtualscroller-content", "ui-widget-content"], ["content", ""], [1, "ui-virtualscroller-list"], [3, "ngStyle", "itemSize", "scrolledIndexChange"], ["viewport", ""], [4, "cdkVirtualFor", "cdkVirtualForOf", "cdkVirtualForTrackBy"], ["class", "ui-virtualscroller-footer ui-widget-header ui-corner-bottom", 4, "ngIf"], [1, "ui-virtualscroller-header", "ui-widget-header", "ui-corner-top"], [3, "ngStyle"], [4, "ngTemplateOutlet", "ngTemplateOutletContext"], [1, "ui-virtualscroller-footer", "ui-widget-header", "ui-corner-bottom"]], template: function VirtualScroller_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c3);
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, VirtualScroller_div_1_Template, 2, 0, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "div", 2, 3);
        ɵngcc0.ɵɵelementStart(4, "ul", 4);
        ɵngcc0.ɵɵelementStart(5, "cdk-virtual-scroll-viewport", 5, 6);
        ɵngcc0.ɵɵlistener("scrolledIndexChange", function VirtualScroller_Template_cdk_virtual_scroll_viewport_scrolledIndexChange_5_listener($event) { return ctx.onScrollIndexChange($event); });
        ɵngcc0.ɵɵtemplate(7, VirtualScroller_ng_container_7_Template, 3, 13, "ng-container", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(8, VirtualScroller_div_8_Template, 2, 0, "div", 8);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.styleClass);
        ɵngcc0.ɵɵproperty("ngClass", "ui-virtualscroller ui-widget")("ngStyle", ctx.style);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.header);
        ɵngcc0.ɵɵadvance(4);
        ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction1(10, _c1, ctx.scrollHeight))("itemSize", ctx.itemSize);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("cdkVirtualForOf", ctx.value)("cdkVirtualForTrackBy", ctx.trackBy);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.footer);
    } }, directives: [ɵngcc1.NgClass, ɵngcc1.NgStyle, ɵngcc1.NgIf, ɵngcc2.CdkVirtualScrollViewport, ɵngcc2.CdkFixedSizeVirtualScroll, ɵngcc2.CdkVirtualForOf, ɵngcc1.NgTemplateOutlet], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(VirtualScroller, [{
        type: core.Component,
        args: [{
                selector: 'p-virtualScroller',
                template: "\n        <div [ngClass]=\"'ui-virtualscroller ui-widget'\" [ngStyle]=\"style\" [class]=\"styleClass\">\n            <div class=\"ui-virtualscroller-header ui-widget-header ui-corner-top\" *ngIf=\"header\">\n                <ng-content select=\"p-header\"></ng-content>\n            </div>\n            <div #content class=\"ui-virtualscroller-content ui-widget-content\">\n                <ul class=\"ui-virtualscroller-list\">\n                    <cdk-virtual-scroll-viewport #viewport [ngStyle]=\"{'height': scrollHeight}\" [itemSize]=\"itemSize\" (scrolledIndexChange)=\"onScrollIndexChange($event)\">\n                        <ng-container *cdkVirtualFor=\"let item of value; trackBy: trackBy; let i = index; let c = count; let f = first; let l = last; let e = even; let o = odd; \">\n                            <li [ngStyle]=\"{'height': itemSize + 'px'}\">\n                                <ng-container *ngTemplateOutlet=\"item ? itemTemplate : loadingItemTemplate; context: {$implicit: item, index: i, count: c, first: f, last: l, even: e, odd: o}\"></ng-container>\n                            </li>\n                        </ng-container>\n                    </cdk-virtual-scroll-viewport>\n                </ul>\n            </div>\n            <div class=\"ui-virtualscroller-footer ui-widget-header ui-corner-bottom\" *ngIf=\"footer\">\n                <ng-content select=\"p-footer\"></ng-content>\n            </div>\n        </div>\n    "
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }]; }, { cache: [{
            type: core.Input
        }], first: [{
            type: core.Input
        }], trackBy: [{
            type: core.Input
        }], onLazyLoad: [{
            type: core.Output
        }], totalRecords: [{
            type: core.Input
        }], value: [{
            type: core.Input
        }], itemSize: [{
            type: core.Input
        }], style: [{
            type: core.Input
        }], styleClass: [{
            type: core.Input
        }], scrollHeight: [{
            type: core.Input
        }], lazy: [{
            type: core.Input
        }], rows: [{
            type: core.Input
        }], header: [{
            type: core.ContentChild,
            args: [api.Header, { static: true }]
        }], footer: [{
            type: core.ContentChild,
            args: [api.Footer, { static: true }]
        }], templates: [{
            type: core.ContentChildren,
            args: [api.PrimeTemplate]
        }], viewPortViewChild: [{
            type: core.ViewChild,
            args: ['viewport', { static: true }]
        }] }); })();
        return VirtualScroller;
    }());
    var VirtualScrollerModule = /** @class */ (function () {
        function VirtualScrollerModule() {
        }
VirtualScrollerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: VirtualScrollerModule });
VirtualScrollerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function VirtualScrollerModule_Factory(t) { return new (t || VirtualScrollerModule)(); }, imports: [[common.CommonModule, scrolling.ScrollingModule], ɵngcc3.SharedModule, ɵngcc2.ScrollingModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(VirtualScrollerModule, { declarations: [VirtualScroller], imports: [ɵngcc1.CommonModule, ɵngcc2.ScrollingModule], exports: [VirtualScroller, ɵngcc3.SharedModule, ɵngcc2.ScrollingModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(VirtualScrollerModule, [{
        type: core.NgModule,
        args: [{
                imports: [common.CommonModule, scrolling.ScrollingModule],
                exports: [VirtualScroller, api.SharedModule, scrolling.ScrollingModule],
                declarations: [VirtualScroller]
            }]
    }], function () { return []; }, null); })();
        return VirtualScrollerModule;
    }());

    exports.VirtualScroller = VirtualScroller;
    exports.VirtualScrollerModule = VirtualScrollerModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
