// primeng/bundles/primeng-megamenu.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('@angular/core'),require('@angular/common'),require('@angular/router'),exports, require('@angular/core'), require('@angular/common'), require('primeng/dom'), require('@angular/router')) :
    typeof define === 'function' && define.amd ? define('primeng/megamenu', ['@angular/core','@angular/common','@angular/router','exports', '@angular/core', '@angular/common', 'primeng/dom', '@angular/router'], factory) :
    (global = global || self, factory(global.ng.core,global.ng.common,global.ng.router,(global.primeng = global.primeng || {}, global.primeng.megamenu = {}), global.ng.core, global.ng.common, global.primeng.dom, global.ng.router));
}(this, (function (ɵngcc0,ɵngcc1,ɵngcc2,exports, core, common, dom, router) { 
var _c0 = function (a0) { return { "ui-helper-hidden": a0 }; };
function MegaMenu_ng_template_2_li_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 6);
} if (rf & 2) {
    var category_r332 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(1, _c0, category_r332.visible === false));
} }
function MegaMenu_ng_template_2_li_1_a_2_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 16);
} if (rf & 2) {
    var category_r332 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵproperty("ngClass", category_r332.icon);
} }
var _c1 = function (a0, a1) { return { "pi-caret-down": a0, "pi-caret-right": a1 }; };
function MegaMenu_ng_template_2_li_1_a_2_span_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 17);
} if (rf & 2) {
    var ctx_r341 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(1, _c1, ctx_r341.orientation == "horizontal", ctx_r341.orientation == "vertical"));
} }
var _c2 = function (a1) { return { "ui-menuitem-link ui-corner-all": true, "ui-state-disabled": a1 }; };
function MegaMenu_ng_template_2_li_1_a_2_Template(rf, ctx) { if (rf & 1) {
    var _r344 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 12);
    ɵngcc0.ɵɵlistener("click", function MegaMenu_ng_template_2_li_1_a_2_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r344); var category_r332 = ɵngcc0.ɵɵnextContext(2).$implicit; var ctx_r343 = ɵngcc0.ɵɵnextContext(); return ctx_r343.itemClick($event, category_r332); });
    ɵngcc0.ɵɵtemplate(1, MegaMenu_ng_template_2_li_1_a_2_span_1_Template, 1, 1, "span", 13);
    ɵngcc0.ɵɵelementStart(2, "span", 14);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(4, MegaMenu_ng_template_2_li_1_a_2_span_4_Template, 1, 4, "span", 15);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var category_r332 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵclassMap(category_r332.styleClass);
    ɵngcc0.ɵɵproperty("href", category_r332.url || "#", ɵngcc0.ɵɵsanitizeUrl)("ngClass", ɵngcc0.ɵɵpureFunction1(12, _c2, category_r332.disabled))("ngStyle", category_r332.style);
    ɵngcc0.ɵɵattribute("target", category_r332.target)("title", category_r332.title)("id", category_r332.id)("tabindex", category_r332.tabindex ? category_r332.tabindex : "0");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", category_r332.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(category_r332.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", category_r332.items);
} }
function MegaMenu_ng_template_2_li_1_a_3_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 16);
} if (rf & 2) {
    var category_r332 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵproperty("ngClass", category_r332.icon);
} }
var _c3 = function () { return { exact: false }; };
function MegaMenu_ng_template_2_li_1_a_3_Template(rf, ctx) { if (rf & 1) {
    var _r350 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 18);
    ɵngcc0.ɵɵlistener("click", function MegaMenu_ng_template_2_li_1_a_3_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r350); var category_r332 = ɵngcc0.ɵɵnextContext(2).$implicit; var ctx_r349 = ɵngcc0.ɵɵnextContext(); return ctx_r349.itemClick($event, category_r332); });
    ɵngcc0.ɵɵtemplate(1, MegaMenu_ng_template_2_li_1_a_3_span_1_Template, 1, 1, "span", 13);
    ɵngcc0.ɵɵelementStart(2, "span", 14);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var category_r332 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵclassMap(category_r332.styleClass);
    ɵngcc0.ɵɵproperty("routerLink", category_r332.routerLink)("queryParams", category_r332.queryParams)("routerLinkActive", "ui-state-active")("routerLinkActiveOptions", category_r332.routerLinkActiveOptions || ɵngcc0.ɵɵpureFunction0(14, _c3))("ngClass", ɵngcc0.ɵɵpureFunction1(15, _c2, category_r332.disabled))("ngStyle", category_r332.style);
    ɵngcc0.ɵɵattribute("tabindex", category_r332.tabindex ? category_r332.tabindex : "0")("target", category_r332.target)("title", category_r332.title)("id", category_r332.id);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", category_r332.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(category_r332.label);
} }
function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "li", 25);
} if (rf & 2) {
    var item_r358 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(1, _c0, item_r358.visible === false));
} }
function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_a_1_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 16);
} if (rf & 2) {
    var item_r358 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵproperty("ngClass", item_r358.icon);
} }
var _c4 = function (a0) { return { "ui-state-disabled": a0 }; };
function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_a_1_Template(rf, ctx) { if (rf & 1) {
    var _r367 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 29);
    ɵngcc0.ɵɵlistener("click", function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_a_1_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r367); var item_r358 = ɵngcc0.ɵɵnextContext(2).$implicit; var ctx_r366 = ɵngcc0.ɵɵnextContext(6); return ctx_r366.itemClick($event, item_r358); });
    ɵngcc0.ɵɵtemplate(1, MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_a_1_span_1_Template, 1, 1, "span", 13);
    ɵngcc0.ɵɵelementStart(2, "span", 14);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r358 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("href", item_r358.url || "#", ɵngcc0.ɵɵsanitizeUrl)("ngClass", ɵngcc0.ɵɵpureFunction1(8, _c4, item_r358.disabled));
    ɵngcc0.ɵɵattribute("target", item_r358.target)("title", item_r358.title)("id", item_r358.id)("tabindex", item_r358.tabindex ? item_r358.tabindex : "0");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r358.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r358.label);
} }
function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_a_2_span_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "span", 16);
} if (rf & 2) {
    var item_r358 = ɵngcc0.ɵɵnextContext(3).$implicit;
    ɵngcc0.ɵɵproperty("ngClass", item_r358.icon);
} }
function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_a_2_Template(rf, ctx) { if (rf & 1) {
    var _r373 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "a", 30);
    ɵngcc0.ɵɵlistener("click", function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_a_2_Template_a_click_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r373); var item_r358 = ɵngcc0.ɵɵnextContext(2).$implicit; var ctx_r372 = ɵngcc0.ɵɵnextContext(6); return ctx_r372.itemClick($event, item_r358); });
    ɵngcc0.ɵɵtemplate(1, MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_a_2_span_1_Template, 1, 1, "span", 13);
    ɵngcc0.ɵɵelementStart(2, "span", 14);
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r358 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵproperty("routerLink", item_r358.routerLink)("queryParams", item_r358.queryParams)("routerLinkActive", "ui-state-active")("routerLinkActiveOptions", item_r358.routerLinkActiveOptions || ɵngcc0.ɵɵpureFunction0(11, _c3))("ngClass", ɵngcc0.ɵɵpureFunction1(12, _c4, item_r358.disabled));
    ɵngcc0.ɵɵattribute("tabindex", item_r358.tabindex ? item_r358.tabindex : "0")("target", item_r358.target)("title", item_r358.title)("id", item_r358.id);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r358.icon);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(item_r358.label);
} }
function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 26);
    ɵngcc0.ɵɵtemplate(1, MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_a_1_Template, 4, 10, "a", 27);
    ɵngcc0.ɵɵtemplate(2, MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_a_2_Template, 4, 14, "a", 28);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var item_r358 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(3, _c0, item_r358.visible === false));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !item_r358.routerLink);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r358.routerLink);
} }
function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_0_Template, 1, 3, "li", 23);
    ɵngcc0.ɵɵtemplate(1, MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_li_1_Template, 3, 5, "li", 24);
} if (rf & 2) {
    var item_r358 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngIf", item_r358.separator);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !item_r358.separator);
} }
function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "ul", 21);
    ɵngcc0.ɵɵelementStart(1, "li", 22);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵtemplate(3, MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_ng_template_3_Template, 2, 2, "ng-template", 2);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var submenu_r356 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(submenu_r356.label);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", submenu_r356.items);
} }
function MegaMenu_ng_template_2_li_1_div_4_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div");
    ɵngcc0.ɵɵtemplate(1, MegaMenu_ng_template_2_li_1_div_4_ng_template_2_ng_template_1_Template, 4, 2, "ng-template", 2);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var column_r354 = ctx.$implicit;
    var category_r332 = ɵngcc0.ɵɵnextContext(3).$implicit;
    var ctx_r353 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassMap(ctx_r353.getColumnClass(category_r332));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", column_r354);
} }
function MegaMenu_ng_template_2_li_1_div_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 19);
    ɵngcc0.ɵɵelementStart(1, "div", 20);
    ɵngcc0.ɵɵtemplate(2, MegaMenu_ng_template_2_li_1_div_4_ng_template_2_Template, 2, 3, "ng-template", 2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var category_r332 = ɵngcc0.ɵɵnextContext(2).$implicit;
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngForOf", category_r332.items);
} }
var _c5 = function (a1, a2) { return { "ui-menuitem ui-corner-all": true, "ui-menuitem-active": a1, "ui-helper-hidden": a2 }; };
function MegaMenu_ng_template_2_li_1_Template(rf, ctx) { if (rf & 1) {
    var _r380 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "li", 7, 8);
    ɵngcc0.ɵɵlistener("mouseenter", function MegaMenu_ng_template_2_li_1_Template_li_mouseenter_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r380); var _r336 = ɵngcc0.ɵɵreference(1); var category_r332 = ɵngcc0.ɵɵnextContext().$implicit; var ctx_r379 = ɵngcc0.ɵɵnextContext(); return ctx_r379.onItemMouseEnter($event, _r336, category_r332); })("mouseleave", function MegaMenu_ng_template_2_li_1_Template_li_mouseleave_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r380); var _r336 = ɵngcc0.ɵɵreference(1); var ctx_r382 = ɵngcc0.ɵɵnextContext(2); return ctx_r382.onItemMouseLeave($event, _r336); });
    ɵngcc0.ɵɵtemplate(2, MegaMenu_ng_template_2_li_1_a_2_Template, 5, 14, "a", 9);
    ɵngcc0.ɵɵtemplate(3, MegaMenu_ng_template_2_li_1_a_3_Template, 4, 17, "a", 10);
    ɵngcc0.ɵɵtemplate(4, MegaMenu_ng_template_2_li_1_div_4_Template, 3, 1, "div", 11);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var _r336 = ɵngcc0.ɵɵreference(1);
    var category_r332 = ɵngcc0.ɵɵnextContext().$implicit;
    var ctx_r334 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(4, _c5, _r336 == ctx_r334.activeItem, category_r332.visible === false));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngIf", !category_r332.routerLink);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", category_r332.routerLink);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", category_r332.items);
} }
function MegaMenu_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtemplate(0, MegaMenu_ng_template_2_li_0_Template, 1, 3, "li", 4);
    ɵngcc0.ɵɵtemplate(1, MegaMenu_ng_template_2_li_1_Template, 5, 7, "li", 5);
} if (rf & 2) {
    var category_r332 = ctx.$implicit;
    ɵngcc0.ɵɵproperty("ngIf", category_r332.separator);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", !category_r332.separator);
} }
function MegaMenu_li_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "li", 31);
    ɵngcc0.ɵɵprojection(1);
    ɵngcc0.ɵɵelementEnd();
} }
var _c6 = function (a1, a2) { return { "ui-megamenu ui-widget ui-widget-content ui-corner-all": true, "ui-megamenu-horizontal": a1, "ui-megamenu-vertical": a2 }; };
var _c7 = ["*"];
'use strict';

    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var MegaMenu = /** @class */ (function () {
        function MegaMenu(el, renderer) {
            this.el = el;
            this.renderer = renderer;
            this.orientation = 'horizontal';
            this.autoZIndex = true;
            this.baseZIndex = 0;
        }
        MegaMenu.prototype.onItemMouseEnter = function (event, item, menuitem) {
            if (menuitem.disabled) {
                return;
            }
            if (this.hideTimeout) {
                clearTimeout(this.hideTimeout);
                this.hideTimeout = null;
            }
            this.activeItem = item;
            if (menuitem.items) {
                var submenu = item.children[0].nextElementSibling;
                if (submenu) {
                    if (this.autoZIndex) {
                        submenu.style.zIndex = String(this.baseZIndex + (++dom.DomHandler.zindex));
                    }
                    if (this.orientation === 'horizontal') {
                        submenu.style.top = dom.DomHandler.getOuterHeight(item.children[0]) + 'px';
                        submenu.style.left = '0px';
                    }
                    else if (this.orientation === 'vertical') {
                        submenu.style.top = '0px';
                        submenu.style.left = dom.DomHandler.getOuterWidth(item.children[0]) + 'px';
                    }
                }
            }
        };
        MegaMenu.prototype.onItemMouseLeave = function (event, link) {
            var _this = this;
            this.hideTimeout = setTimeout(function () {
                _this.activeItem = null;
            }, 1000);
        };
        MegaMenu.prototype.itemClick = function (event, item) {
            if (item.disabled) {
                event.preventDefault();
                return;
            }
            if (!item.url) {
                event.preventDefault();
            }
            if (item.command) {
                item.command({
                    originalEvent: event,
                    item: item
                });
            }
            this.activeItem = null;
        };
        MegaMenu.prototype.getColumnClass = function (menuitem) {
            var length = menuitem.items ? menuitem.items.length : 0;
            var columnClass;
            switch (length) {
                case 2:
                    columnClass = 'ui-g-6';
                    break;
                case 3:
                    columnClass = 'ui-g-4';
                    break;
                case 4:
                    columnClass = 'ui-g-3';
                    break;
                case 6:
                    columnClass = 'ui-g-2';
                    break;
                default:
                    columnClass = 'ui-g-12';
                    break;
            }
            return columnClass;
        };
        MegaMenu.ctorParameters = function () { return [
            { type: core.ElementRef },
            { type: core.Renderer2 }
        ]; };
        __decorate([
            core.Input()
        ], MegaMenu.prototype, "model", void 0);
        __decorate([
            core.Input()
        ], MegaMenu.prototype, "style", void 0);
        __decorate([
            core.Input()
        ], MegaMenu.prototype, "styleClass", void 0);
        __decorate([
            core.Input()
        ], MegaMenu.prototype, "orientation", void 0);
        __decorate([
            core.Input()
        ], MegaMenu.prototype, "autoZIndex", void 0);
        __decorate([
            core.Input()
        ], MegaMenu.prototype, "baseZIndex", void 0);
MegaMenu.ɵfac = function MegaMenu_Factory(t) { return new (t || MegaMenu)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
MegaMenu.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MegaMenu, selectors: [["p-megaMenu"]], inputs: { orientation: "orientation", autoZIndex: "autoZIndex", baseZIndex: "baseZIndex", model: "model", style: "style", styleClass: "styleClass" }, ngContentSelectors: _c7, decls: 4, vars: 9, consts: [[3, "ngStyle", "ngClass"], ["role", "menubar", 1, "ui-megamenu-root-list"], ["ngFor", "", 3, "ngForOf"], ["class", "ui-menuitem ui-menuitem-custom ui-corner-all", 4, "ngIf"], ["class", "ui-menu-separator ui-widget-content", 3, "ngClass", 4, "ngIf"], [3, "ngClass", "mouseenter", "mouseleave", 4, "ngIf"], [1, "ui-menu-separator", "ui-widget-content", 3, "ngClass"], [3, "ngClass", "mouseenter", "mouseleave"], ["item", ""], [3, "href", "ngClass", "ngStyle", "class", "click", 4, "ngIf"], [3, "routerLink", "queryParams", "routerLinkActive", "routerLinkActiveOptions", "ngClass", "ngStyle", "class", "click", 4, "ngIf"], ["class", "ui-megamenu-panel ui-widget-content ui-corner-all ui-shadow", 4, "ngIf"], [3, "href", "ngClass", "ngStyle", "click"], ["class", "ui-menuitem-icon", 3, "ngClass", 4, "ngIf"], [1, "ui-menuitem-text"], ["class", "ui-submenu-icon pi pi-fw", 3, "ngClass", 4, "ngIf"], [1, "ui-menuitem-icon", 3, "ngClass"], [1, "ui-submenu-icon", "pi", "pi-fw", 3, "ngClass"], [3, "routerLink", "queryParams", "routerLinkActive", "routerLinkActiveOptions", "ngClass", "ngStyle", "click"], [1, "ui-megamenu-panel", "ui-widget-content", "ui-corner-all", "ui-shadow"], [1, "ui-g"], ["role", "menu", 1, "ui-megamenu-submenu"], [1, "ui-widget-header", "ui-megamenu-submenu-header", "ui-corner-all"], ["class", "ui-menu-separator ui-widget-content", "role", "separator", 3, "ngClass", 4, "ngIf"], ["class", "ui-menuitem ui-corner-all", "role", "none", 3, "ngClass", 4, "ngIf"], ["role", "separator", 1, "ui-menu-separator", "ui-widget-content", 3, "ngClass"], ["role", "none", 1, "ui-menuitem", "ui-corner-all", 3, "ngClass"], ["role", "menuitem", "class", "ui-menuitem-link ui-corner-all", 3, "href", "ngClass", "click", 4, "ngIf"], ["role", "menuitem", "class", "ui-menuitem-link ui-corner-all", 3, "routerLink", "queryParams", "routerLinkActive", "routerLinkActiveOptions", "ngClass", "click", 4, "ngIf"], ["role", "menuitem", 1, "ui-menuitem-link", "ui-corner-all", 3, "href", "ngClass", "click"], ["role", "menuitem", 1, "ui-menuitem-link", "ui-corner-all", 3, "routerLink", "queryParams", "routerLinkActive", "routerLinkActiveOptions", "ngClass", "click"], [1, "ui-menuitem", "ui-menuitem-custom", "ui-corner-all"]], template: function MegaMenu_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "ul", 1);
        ɵngcc0.ɵɵtemplate(2, MegaMenu_ng_template_2_Template, 2, 2, "ng-template", 2);
        ɵngcc0.ɵɵtemplate(3, MegaMenu_li_3_Template, 2, 0, "li", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.styleClass);
        ɵngcc0.ɵɵproperty("ngStyle", ctx.style)("ngClass", ɵngcc0.ɵɵpureFunction2(6, _c6, ctx.orientation == "horizontal", ctx.orientation == "vertical"));
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.model);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.orientation === "horizontal");
    } }, directives: [ɵngcc1.NgStyle, ɵngcc1.NgClass, ɵngcc1.NgForOf, ɵngcc1.NgIf, ɵngcc2.RouterLinkWithHref, ɵngcc2.RouterLinkActive], encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MegaMenu, [{
        type: core.Component,
        args: [{
                selector: 'p-megaMenu',
                template: "\n        <div [class]=\"styleClass\" [ngStyle]=\"style\"\n            [ngClass]=\"{'ui-megamenu ui-widget ui-widget-content ui-corner-all':true,'ui-megamenu-horizontal': orientation == 'horizontal','ui-megamenu-vertical': orientation == 'vertical'}\">\n            <ul class=\"ui-megamenu-root-list\" role=\"menubar\">\n                <ng-template ngFor let-category [ngForOf]=\"model\">\n                    <li *ngIf=\"category.separator\" class=\"ui-menu-separator ui-widget-content\" [ngClass]=\"{'ui-helper-hidden': category.visible === false}\">\n                    <li *ngIf=\"!category.separator\" #item [ngClass]=\"{'ui-menuitem ui-corner-all':true,'ui-menuitem-active':item==activeItem, 'ui-helper-hidden': category.visible === false}\"\n                        (mouseenter)=\"onItemMouseEnter($event, item, category)\" (mouseleave)=\"onItemMouseLeave($event, item)\">\n   \n                        <a *ngIf=\"!category.routerLink\" [href]=\"category.url||'#'\" [attr.target]=\"category.target\" [attr.title]=\"category.title\" [attr.id]=\"category.id\" (click)=\"itemClick($event, category)\" [attr.tabindex]=\"category.tabindex ? category.tabindex : '0'\"\n                            [ngClass]=\"{'ui-menuitem-link ui-corner-all':true,'ui-state-disabled':category.disabled}\" [ngStyle]=\"category.style\" [class]=\"category.styleClass\">\n                            <span class=\"ui-menuitem-icon\" *ngIf=\"category.icon\" [ngClass]=\"category.icon\"></span>\n                            <span class=\"ui-menuitem-text\">{{category.label}}</span>\n                            <span *ngIf=\"category.items\" class=\"ui-submenu-icon pi pi-fw\" [ngClass]=\"{'pi-caret-down':orientation=='horizontal','pi-caret-right':orientation=='vertical'}\"></span>\n                        </a>\n                        <a *ngIf=\"category.routerLink\" [routerLink]=\"category.routerLink\" [queryParams]=\"category.queryParams\" [routerLinkActive]=\"'ui-state-active'\" [routerLinkActiveOptions]=\"category.routerLinkActiveOptions||{exact:false}\" [attr.tabindex]=\"category.tabindex ? category.tabindex : '0'\" \n                            [attr.target]=\"category.target\" [attr.title]=\"category.title\" [attr.id]=\"category.id\"\n                            (click)=\"itemClick($event, category)\" [ngClass]=\"{'ui-menuitem-link ui-corner-all':true,'ui-state-disabled':category.disabled}\" [ngStyle]=\"category.style\" [class]=\"category.styleClass\">\n                            <span class=\"ui-menuitem-icon\" *ngIf=\"category.icon\" [ngClass]=\"category.icon\"></span>\n                            <span class=\"ui-menuitem-text\">{{category.label}}</span>\n                        </a>\n\n                        <div class=\"ui-megamenu-panel ui-widget-content ui-corner-all ui-shadow\" *ngIf=\"category.items\">\n                            <div class=\"ui-g\">\n                                <ng-template ngFor let-column [ngForOf]=\"category.items\">\n                                    <div [class]=\"getColumnClass(category)\">\n                                        <ng-template ngFor let-submenu [ngForOf]=\"column\">\n                                            <ul class=\"ui-megamenu-submenu\" role=\"menu\">\n                                                <li class=\"ui-widget-header ui-megamenu-submenu-header ui-corner-all\">{{submenu.label}}</li>\n                                                <ng-template ngFor let-item [ngForOf]=\"submenu.items\">\n                                                    <li *ngIf=\"item.separator\" class=\"ui-menu-separator ui-widget-content\" [ngClass]=\"{'ui-helper-hidden': item.visible === false}\" role=\"separator\">\n                                                    <li *ngIf=\"!item.separator\" class=\"ui-menuitem ui-corner-all\" [ngClass]=\"{'ui-helper-hidden': item.visible === false}\" role=\"none\">\n                                                        <a *ngIf=\"!item.routerLink\" role=\"menuitem\" [href]=\"item.url||'#'\" class=\"ui-menuitem-link ui-corner-all\" [attr.target]=\"item.target\" [attr.title]=\"item.title\" [attr.id]=\"item.id\" [attr.tabindex]=\"item.tabindex ? item.tabindex : '0'\"\n                                                            [ngClass]=\"{'ui-state-disabled':item.disabled}\" (click)=\"itemClick($event, item)\">\n                                                            <span class=\"ui-menuitem-icon\" *ngIf=\"item.icon\" [ngClass]=\"item.icon\"></span>\n                                                            <span class=\"ui-menuitem-text\">{{item.label}}</span>\n                                                        </a>\n                                                        <a *ngIf=\"item.routerLink\" role=\"menuitem\" [routerLink]=\"item.routerLink\" [queryParams]=\"item.queryParams\" [routerLinkActive]=\"'ui-state-active'\" [attr.tabindex]=\"item.tabindex ? item.tabindex : '0'\"\n                                                            [routerLinkActiveOptions]=\"item.routerLinkActiveOptions||{exact:false}\" class=\"ui-menuitem-link ui-corner-all\" \n                                                             [attr.target]=\"item.target\" [attr.title]=\"item.title\" [attr.id]=\"item.id\"\n                                                            [ngClass]=\"{'ui-state-disabled':item.disabled}\" (click)=\"itemClick($event, item)\">\n                                                            <span class=\"ui-menuitem-icon\" *ngIf=\"item.icon\" [ngClass]=\"item.icon\"></span>\n                                                            <span class=\"ui-menuitem-text\">{{item.label}}</span>\n                                                        </a>\n                                                    </li>\n                                                </ng-template>\n                                            </ul>\n                                        </ng-template>\n                                    </div>\n                                </ng-template>\n                            </div>\n                        </div>\n                    </li>\n                </ng-template>\n                <li class=\"ui-menuitem ui-menuitem-custom ui-corner-all\" *ngIf=\"orientation === 'horizontal'\">\n                    <ng-content></ng-content>\n                </li>\n            </ul>\n        </div>\n    "
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { orientation: [{
            type: core.Input
        }], autoZIndex: [{
            type: core.Input
        }], baseZIndex: [{
            type: core.Input
        }], model: [{
            type: core.Input
        }], style: [{
            type: core.Input
        }], styleClass: [{
            type: core.Input
        }] }); })();
        return MegaMenu;
    }());
    var MegaMenuModule = /** @class */ (function () {
        function MegaMenuModule() {
        }
MegaMenuModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MegaMenuModule });
MegaMenuModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MegaMenuModule_Factory(t) { return new (t || MegaMenuModule)(); }, imports: [[common.CommonModule, router.RouterModule], ɵngcc2.RouterModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MegaMenuModule, { declarations: [MegaMenu], imports: [ɵngcc1.CommonModule, ɵngcc2.RouterModule], exports: [MegaMenu, ɵngcc2.RouterModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MegaMenuModule, [{
        type: core.NgModule,
        args: [{
                imports: [common.CommonModule, router.RouterModule],
                exports: [MegaMenu, router.RouterModule],
                declarations: [MegaMenu]
            }]
    }], function () { return []; }, null); })();
        return MegaMenuModule;
    }());

    exports.MegaMenu = MegaMenu;
    exports.MegaMenuModule = MegaMenuModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
